File: Day10/day10_part1.hs
import Data.List

main :: IO ()
main = do input <- readFile "input.txt"
          print . mult . parse . (0:) . sort . (map (\ x -> read x :: Int)) . words $ input
          
parse :: [Int] -> (Int, Int)
parse (x1:[]) = (0,1)
parse (x1:x2:xs) = ( \(a,b) -> if x2 - x1 == 1 then (a+1, b) else (a, b+1)) (parse (x2:xs))

mult :: (Int, Int) -> Int
mult (a,b) = a*b

File: 2022_AoC/day10/day10_part1.rs
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let data = std::fs::read_to_string("input.txt")?;
    let input: Vec<Instr> = data.lines()
                                .map(|x| parse(x))
                                .collect();

    let mut xs: Vec<i32> = vec![1, 1];
    execute(&input, &mut xs);
    let result = get_result(&xs);
    println!("{:?}", result); 

    Ok(())
}

#[derive(Debug)]
enum Instr {
    Noop, 
    Addx(i32),
}

fn parse(line: &str) -> Instr {
    let s: Vec<&str> = line.split_whitespace().collect();
    match s[0] {
        "noop" => Instr::Noop,
        "addx" => Instr::Addx(s[1].parse().unwrap()),
        &_     => panic!("Unexpected instruction"),
    }
}

fn execute(instrs: &Vec<Instr>, xs: &mut Vec<i32>) {
    for instr in instrs.iter() {
        let x = xs.last().unwrap().clone();
        match instr {
            Instr::Noop     => xs.push(x),
            Instr::Addx(nb) => {
                xs.push(x);
                xs.push(x+nb);
            }
        }
    }
}

fn get_result(xs: &Vec<i32>) -> i32 {
    let mut result: i32 = 0;
    for i in [20, 60, 100, 140, 180, 220].iter() {
        result += xs[*i as usize] * i;
    }
    result
}
